File: 94.binary-tree-inorder-traversal.java
/*
 * @lc app=leetcode id=94 lang=java
 *
 * [94] Binary Tree Inorder Traversal
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        if (root == null) return res;
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            TreeNode node = stack.pop();
            if (node.left == null && node.right == null) {
                res.add(node.val);
            } else {
                if (node.right != null) stack.push(node.right);
                stack.push(new TreeNode(node.val));
                if (node.left != null) stack.push(node.left);
            }
        }
        return res;
    }
}

File: go/94.binary-tree-inorder-traversal.go
/*
 * @lc app=leetcode id=94 lang=golang
 *
 * [94] Binary Tree Inorder Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
    if root == nil {
		return nil
	}
	res := make([]int, 0)
	queue := []*TreeNode{root}
	for len(queue) != 0 {
		node := queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		if node.Left == nil && node.Right == nil {
			res = append(res, node.Val)
			continue
		}
		right, left := node.Right, node.Left
		node.Right, node.Left = nil, nil
		if right != nil {
			queue = append(queue, right)
		}
		queue = append(queue, node)
		if left != nil {
			queue = append(queue, left)
		}
	}
	return res
}
// @lc code=end

